ding/reward_model/count_based_model.py

```diff
@@ -0,0 +1,354 @@
+        self.tb_logger = tb_logger
+        assert self._counter_type in ['SimHash', 'AutoEncoder', 'PixelCNN']
+        # TODO: copy
+        if self._counter_type == 'PixelCNN':
+            print(cfg)
+            self._counter = GatedPixelCNN(
+                in_channels=cfg.in_channels,
+                out_channels=cfg.out_channels,
+                n_gated=cfg.n_gated,
+                gated_channels=cfg.gated_channels,
+                head_channels=cfg.head_channels,
+                q_level=cfg.q_level,
+            )
+            self._counter.to(self.device)
+            self.intrinsic_reward_type = cfg.intrinsic_reward_type
+            self.obs_shape = (cfg.img_height, cfg.img_width)
+            self.index_range = np.arange(cfg.img_height * cfg.img_width)
+            assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+            self.train_data = []
+            self.opt = optim.RMSprop(
+                self._counter.parameters(),
+                momentum=0.9,
+                weight_decay=0.95,
+                eps=1e-4,
+            )
+
+    def _train(self):
+        train_data: list = self.train_data
+        train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+        loss = nn.CrossEntropyLoss()(flattened_logits, target_pixel_loss.long())
+        loss = loss.mean()
+        self.opt.zero_grad()
+        loss.backward()
+        self.opt.step()
+
+    def train(self) -> None:
+        self._train()
+        # self.clear_data()
+
+    def estimate(self, data, t) -> None:
+        """
+            # TODO:
+        """
+        if self._counter_type == 'PixelCNN':
+            obs = self._collect_states(data)
+            obs = torch.stack(obs).to(self.device)
+
+            prob = (self._probs(obs) + 1e-8).sum().item()
+
+            with torch.no_grad():
+                recoding_prob = (self._probs(obs) + 1e-8).sum().item()
+
+            pred_gain = np.log(recoding_prob) - np.log(prob)
+
+            intrinsic_reward = [pow(
+                exp(self.cfg.bonus_coeffient * pow(t+1, -0.5) * pred_gain) - 1, 0.5
+            )]
+
+            for item, rew in zip(data, intrinsic_reward):
+                if self.intrinsic_reward_type == 'add':
+                    item['reward'] += rew
+                elif self.intrinsic_reward_type == 'new':
+                    item['intrinsic_reward'] = rew
+                elif self.intrinsic_reward_type == 'assign':
+                    item['reward'] = rew
+
+    def _probs(self, obs):
+        _, indexes, target = self._counter(obs)
+        pred_prob = target[self.index_range, indexes.long()]
+        return pred_prob
+
+    def _collect_states(self, data):
+        obs = []
+        for item in data:
+            state = item['obs']
+            state = state.permute(1, 2, 0)
+            state = state.numpy()
+            state = cv2.resize(state, self.obs_shape)
```

Comment:
Use `torch.nn.functional.interpolate` to replace `cv2.resize`, so you don't need transform to numpy